WalletWasabi/WabiSabi/Client/SmartRequestNode.cs

```diff
@@ -0,0 +1,170 @@
+			vsizeCredentials = vsizeCredentials.Concat(Enumerable.Range(0, VsizeCredentialTasks.Count() - vsizeCredentials.Count()).Select(_ => ZeroVsizeCredentialPool.GetZeroCredential()));
+
+			foreach ((TaskCompletionSource<Credential> tcs, Credential credential) in AmountCredentialTasks.Zip(amountCredentials))
+			{
+				tcs.SetResult(credential);
+			}
+			foreach ((TaskCompletionSource<Credential> tcs, Credential credential) in VsizeCredentialTasks.Zip(vsizeCredentials))
+			{
+				tcs.SetResult(credential);
+			}
+		}
+
+		public async Task StartReissuanceAsync(BobClient bobClient, IEnumerable<long> amounts, IEnumerable<long> vsizes, CancellationToken cancellationToken)
+		{
+			await Task.WhenAll(AmountCredentialToPresentTasks.Concat(VsizeCredentialToPresentTasks)).ConfigureAwait(false);
+			IEnumerable<Credential> inputAmountCredentials = AmountCredentialToPresentTasks.Select(x => x.Result);
+			IEnumerable<Credential> inputVsizeCredentials = VsizeCredentialToPresentTasks.Select(x => x.Result);
+			var amountsToRequest = AddExtraCredential(amounts, inputAmountCredentials);
+			var vsizesToRequest = AddExtraCredential(vsizes, inputVsizeCredentials);
+
+			(IEnumerable<Credential> RealAmountCredentials, IEnumerable<Credential> RealVsizeCredentials) result = await bobClient.ReissueCredentialsAsync(
+				amountsToRequest,
+				vsizesToRequest,
+				inputAmountCredentials,
+				inputVsizeCredentials,
+				cancellationToken).ConfigureAwait(false);
+
+			// TODO keep the credentials that were not needed by the graph
+			// Strip out any credentials that have to be requested for the
+			// remaining amount but do not have a corresponding edge in the
+			// graph (because no request node depends on them.)
+			var amountCredentials = result.RealAmountCredentials.Take(amounts.Where(v => v != 0).Count());
+			var vsizeCredentials = result.RealVsizeCredentials.Take(vsizes.Where(v => v != 0).Count());
+
+			// TODO remove
+			// Add back any zero credentials that do have a corresponding edge
+			// (and therefore a task completion source), so that the
+			// `amountCredentials` enumerator matches the `amounts` enumerator.
+			// Assumes that `amounts` is ordered by value descending, so zero
+			// credentials are at the end. (same for vsizes).
+			amountCredentials = amountCredentials.Concat(Enumerable.Range(0, AmountCredentialTasks.Count() - amountCredentials.Count()).Select(_ => ZeroAmountCredentialPool.GetZeroCredential()));
+			vsizeCredentials = vsizeCredentials.Concat(Enumerable.Range(0, VsizeCredentialTasks.Count() - vsizeCredentials.Count()).Select(_ => ZeroVsizeCredentialPool.GetZeroCredential()));
+
+			foreach ((TaskCompletionSource<Credential> tcs, Credential credential) in AmountCredentialTasks.Zip(amountCredentials))
+			{
+				tcs.SetResult(credential);
+			}
+			foreach ((TaskCompletionSource<Credential> tcs, Credential credential) in VsizeCredentialTasks.Zip(vsizeCredentials))
+			{
+				tcs.SetResult(credential);
+			}
```

Comment:
I don't want to start a war here but imo so much verbosity makes Cobol look sexier.

```suggestion
			foreach (var (tcs, credential) in AmountCredentialTasks.Zip(amountCredentials))
			{
				tcs.SetResult(credential);
			}
			foreach (var (tcs, credential) in VsizeCredentialTasks.Zip(vsizeCredentials))
			{
				tcs.SetResult(credential);
			}
```

Reply:
It is!